File: packages/ash-node/src/query-url-comprehensive.test.ts
/**
 * Query/URL Encoding Comprehensive Tests
 *
 * Tests for query string and URL handling covering:
 * - Percent encoding (RFC 3986)
 * - Special characters
 * - Unicode handling
 * - Base64URL encoding/decoding
 * - Edge cases and security concerns
 */

import { describe, it, expect, beforeAll } from 'vitest';
import {
  ashInit,
  ashCanonicalizeQuery,
  ashCanonicalizeQueryNative,
  ashCanonicalizeUrlencoded,
  ashCanonicalizeUrlencodedNative,
  ashBase64UrlEncode,
  ashBase64UrlDecode,
  ashBase64UrlDecodeToBuffer,
  ashNormalizeBinding,
  ashNormalizeBindingFromUrl,
} from './index';

// Initialize WASM
beforeAll(() => {
  try {
    ashInit();
  } catch {
    // WASM not available, tests will use native fallback
  }
});

describe('Query/URL Encoding Comprehensive Tests', () => {
  describe('Percent Encoding - RFC 3986', () => {
    it('uppercases lowercase percent-encoded hex', () => {
      expect(ashCanonicalizeQueryNative('a=%2f')).toBe('a=%2F');
      expect(ashCanonicalizeQueryNative('a=%2F')).toBe('a=%2F');
    });

    it('normalizes mixed-case percent encoding', () => {
      expect(ashCanonicalizeQueryNative('a=%2f&b=%2F&c=%2f')).toBe('a=%2F&b=%2F&c=%2F');
    });

    it('handles encoded space (%20)', () => {
      expect(ashCanonicalizeQueryNative('a=hello%20world')).toBe('a=hello%20world');
    });

    it('handles encoded ampersand (%26)', () => {
      expect(ashCanonicalizeQueryNative('a=foo%26bar')).toBe('a=foo%26bar');
    });

    it('handles encoded equals (%3D)', () => {
      expect(ashCanonicalizeQueryNative('a=foo%3Dbar')).toBe('a=foo%3Dbar');
    });

    it('handles all special characters encoded', () => {
      const specialChars = [
        ['%21', '!'],
        ['%23', '#'],
        ['%24', '$'],
        ['%25', '%'],
        ['%26', '&'],
        ['%27', "'"],
        ['%28', '('],
        ['%29', ')'],
        ['%2A', '*'],
        ['%2B', '+'],
        ['%2C', ','],
        ['%2F', '/'],
        ['%3A', ':'],
        ['%3B', ';'],
        ['%3D', '='],
        ['%3F', '?'],
        ['%40', '@'],
        ['%5B', '['],
        ['%5D', ']'],
      ];

      for (const [encoded] of specialChars) {
        const result = ashCanonicalizeQueryNative(`key=${encoded}`);
        expect(result).toContain(encoded.toUpperCase());
      }
    });

    it('preserves unreserved characters', () => {
      // Unreserved: A-Z a-z 0-9 - . _ ~
      const unreserved = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
      expect(ashCanonicalizeQueryNative(`key=${unreserved}`)).toBe(`key=${unreserved}`);
    });
  });

  describe('Special Character Handling', () => {
    it('treats + as literal plus (not space)', () => {
      expect(ashCanonicalizeQueryNative('a+b=1')).toBe('a+b=1');
      expect(ashCanonicalizeQueryNative('a=b+c')).toBe('a=b+c');
    });

    it('distinguishes + from %2B', () => {
      // + is literal plus, %2B is also plus but encoded
      expect(ashCanonicalizeQueryNative('a=%2B')).toBe('a=%2B');
    });

    it('handles equals sign in value', () => {
      expect(ashCanonicalizeQueryNative('key=a=b')).toBe('key=a=b');
      expect(ashCanonicalizeQueryNative('key=a=b=c')).toBe('key=a=b=c');
    });

    it('handles empty value', () => {
      expect(ashCanonicalizeQueryNative('key=')).toBe('key=');
    });

    it('handles key without value (no equals)', () => {
      expect(ashCanonicalizeQueryNative('flag')).toBe('flag=');
    });

    it('handles empty key', () => {
      expect(ashCanonicalizeQueryNative('=value')).toBe('=value');
    });

    it('handles multiple empty values', () => {
      expect(ashCanonicalizeQueryNative('a=&b=&c=')).toBe('a=&b=&c=');
    });

    it('handles leading question mark', () => {
      expect(ashCanonicalizeQueryNative('?a=1&b=2')).toBe('a=1&b=2');
    });

    it('strips fragment', () => {
      expect(ashCanonicalizeQueryNative('a=1#fragment')).toBe('a=1');
      expect(ashCanonicalizeQueryNative('a=1&b=2#section')).toBe('a=1&b=2');
    });

    it('handles fragment with query-like content', () => {
      expect(ashCanonicalizeQueryNative('a=1#b=2&c=3')).toBe('a=1');
    });
  });

  describe('Unicode Handling', () => {
    it('handles percent-encoded Unicode', () => {
      // 日 = %E6%97%A5
      const result = ashCanonicalizeQueryNative('name=%E6%97%A5');
      expect(result).toBe('name=%E6%97%A5');
    });

    it('normalizes lowercase encoded Unicode', () => {
      // Same character with lowercase hex
      expect(ashCanonicalizeQueryNative('name=%e6%97%a5')).toBe('name=%E6%97%A5');
    });

    it('handles emojis (encoded)', () => {
      // 🎉 = %F0%9F%8E%89
      const result = ashCanonicalizeQueryNative('emoji=%F0%9F%8E%89');
      expect(result).toBe('emoji=%F0%9F%8E%89');
    });

    it('handles raw Unicode in query (passed through)', () => {
      // Raw Unicode should be passed through as-is
      const result = ashCanonicalizeQueryNative('name=日本語');
      expect(result).toContain('name=');
    });

    it('handles mixed encoded and raw Unicode', () => {
      const result = ashCanonicalizeQueryNative('a=日本&b=%E6%97%A5');
      expect(result).toContain('a=');
      expect(result).toContain('b=');
    });
  });

  describe('Sorting', () => {
    it('sorts parameters by key', () => {
      expect(ashCanonicalizeQueryNative('z=1&a=2&m=3')).toBe('a=2&m=3&z=1');
    });

    it('sorts duplicate keys by value', () => {
      expect(ashCanonicalizeQueryNative('a=z&a=a&a=m')).toBe('a=a&a=m&a=z');
    });

    it('sorts numeric keys as strings', () => {
      expect(ashCanonicalizeQueryNative('10=a&2=b&1=c')).toBe('1=c&10=a&2=b');
    });

    it('sorts keys with special characters', () => {
      // Byte-wise sorting
      const result = ashCanonicalizeQueryNative('a-b=1&a_c=2&ab=3');
      expect(result).toBe('a-b=1&a_c=2&ab=3');
    });

    it('sorts case-sensitively (uppercase before lowercase)', () => {
      expect(ashCanonicalizeQueryNative('b=1&B=2&a=3&A=4')).toBe('A=4&B=2&a=3&b=1');
    });

    it('maintains order for identical keys with same values', () => {
      // When key and value are same, order should be stable
      const result = ashCanonicalizeQueryNative('a=1&a=1');
      expect(result).toBe('a=1&a=1');
    });
  });

  describe('URL-Encoded Form Data', () => {
    it('sorts form data parameters', () => {
      expect(ashCanonicalizeUrlencodedNative('z=1&a=2')).toBe('a=2&z=1');
    });

    it('handles plus as literal in form data', () => {
      // In form data, + is typically used for space, but we treat it as literal
      // and encode as %2B
      const result = ashCanonicalizeUrlencodedNative('a+b=1');
      expect(result).toBe('a%2Bb=1');
    });

    it('encodes plus in values', () => {
      const result = ashCanonicalizeUrlencodedNative('name=John+Doe');
      expect(result).toBe('name=John%2BDoe');
    });

    it('matches query canonicalization for non-plus content', () => {
      const query = 'a=1&b=2&c=3';
      expect(ashCanonicalizeUrlencodedNative(query)).toBe(ashCanonicalizeQueryNative(query));
    });
  });

  describe('Base64URL Encoding', () => {
    it('encodes string to base64url', () => {
      const result = ashBase64UrlEncode('Hello, World!');
      expect(result).toBe('SGVsbG8sIFdvcmxkIQ');
    });

    it('encodes buffer to base64url', () => {
      const buffer = Buffer.from('test data');
      const result = ashBase64UrlEncode(buffer);
      expect(result).toBe('dGVzdCBkYXRh');
    });

    it('uses URL-safe characters (no +/)', () => {
      // Input that would produce + and / in regular base64
      const input = '>>??';
      const result = ashBase64UrlEncode(input);
      expect(result).not.toContain('+');
      expect(result).not.toContain('/');
    });

    it('does not include padding', () => {
      const result = ashBase64UrlEncode('test');
      expect(result).not.toContain('=');
    });

    it('handles empty input', () => {
      expect(ashBase64UrlEncode('')).toBe('');
    });

    it('handles binary data', () => {
      const binary = Buffer.from([0x00, 0xff, 0x80, 0x7f]);
      const result = ashBase64UrlEncode(binary);
      expect(result).toBeDefined();
      expect(typeof result).toBe('string');
    });
  });

  describe('Base64URL Decoding', () => {
    it('decodes base64url to string', () => {
      const result = ashBase64UrlDecode('SGVsbG8sIFdvcmxkIQ');
      expect(result).toBe('Hello, World!');
    });

    it('decodes base64url to buffer', () => {
      const result = ashBase64UrlDecodeToBuffer('dGVzdCBkYXRh');
      expect(result.toString()).toBe('test data');
    });

    it('handles URL-safe characters', () => {
      // Input with - and _ (URL-safe replacements for + and /)
      const encoded = ashBase64UrlEncode('test>>??data');
      const decoded = ashBase64UrlDecode(encoded);
      expect(decoded).toBe('test>>??data');
    });

    it('handles empty input', () => {
      expect(ashBase64UrlDecode('')).toBe('');
    });

    it('handles binary data round-trip', () => {
      const original = Buffer.from([0x00, 0xff, 0x80, 0x7f]);
      const encoded = ashBase64UrlEncode(original);
      const decoded = ashBase64UrlDecodeToBuffer(encoded);
      expect(decoded.equals(original)).toBe(true);
    });

    it('roundtrip encoding/decoding preserves data', () => {
      const testCases = [
        'Hello, World!',
        '日本語',
        'Special chars: !@#$%^&*()',
        '\x00\x01\x02\x03',
        'a'.repeat(1000),
      ];

      for (const input of testCases) {
        const encoded = ashBase64UrlEncode(input);
        const decoded = ashBase64UrlDecode(encoded);
        expect(decoded).toBe(input);
      }
    });
  });

  describe('Full URL Normalization', () => {
    it('handles URL with query string', () => {
      const result = ashNormalizeBindingFromUrl('GET', '/api/users?page=1&limit=10');
      expect(result).toBe('GET|/api/users|limit=10&page=1');
    });

    it('handles URL without query string', () => {
      const result = ashNormalizeBindingFromUrl('GET', '/api/users');
      expect(result).toBe('GET|/api/users|');
    });

    it('handles URL with fragment', () => {
      const result = ashNormalizeBindingFromUrl('GET', '/api/users?page=1#section');
      expect(result).toBe('GET|/api/users|page=1');
    });

    it('handles URL with multiple question marks', () => {
      // Only first ? is query delimiter
      // WASM may encode ? in query values as %3F
      const result = ashNormalizeBindingFromUrl('GET', '/search?q=what?is&a=1');
      expect(result).toContain('GET|/search|');
      expect(result).toContain('a=1');
      expect(result).toContain('q=what');
    });

    it('handles complex path with query', () => {
      // WASM may encode comma as %2C
      const result = ashNormalizeBindingFromUrl('POST', '/api/v2/../v3/users?fields=name,email&sort=name');
      expect(result).toContain('POST|/api/v3/users|');
      expect(result).toContain('fields=name');
      expect(result).toContain('sort=name');
    });
  });

  describe('Edge Cases', () => {
    it('handles very long query strings', () => {
      const params = Array.from({ length: 100 }, (_, i) => `key${i}=value${i}`).join('&');
      const result = ashCanonicalizeQueryNative(params);
      expect(result.split('&').length).toBe(100);
    });

    it('handles query with only ampersands', () => {
      expect(ashCanonicalizeQueryNative('&&&')).toBe('');
    });

    it('handles consecutive ampersands', () => {
      expect(ashCanonicalizeQueryNative('a=1&&b=2&&&c=3')).toBe('a=1&b=2&c=3');
    });

    it('handles whitespace in query', () => {
      expect(ashCanonicalizeQueryNative('   ')).toBe('');
    });

    it('handles trailing ampersand', () => {
      expect(ashCanonicalizeQueryNative('a=1&')).toBe('a=1');
    });

    it('handles leading ampersand', () => {
      expect(ashCanonicalizeQueryNative('&a=1')).toBe('a=1');
    });

    it('handles null byte in encoded form', () => {
      const result = ashCanonicalizeQueryNative('key=%00');
      expect(result).toBe('key=%00');
    });

    it('handles incomplete percent encoding', () => {
      // Incomplete encoding should be passed through
      const result = ashCanonicalizeQueryNative('key=%2');
      expect(result).toContain('key=');
    });

    it('handles percent without hex digits', () => {
      const result = ashCanonicalizeQueryNative('key=100%');
      expect(result).toContain('key=');
    });
  });

  describe('Security Edge Cases', () => {
    it('handles SQL injection in query values', () => {
      const injection = encodeURIComponent("'; DROP TABLE users; --");
      const result = ashCanonicalizeQueryNative(`id=${injection}`);
      expect(result).toContain('id=');
    });

    it('handles XSS in query values', () => {
      const xss = encodeURIComponent('<script>alert(1)</script>');
      const result = ashCanonicalizeQueryNative(`msg=${xss}`);
      expect(result).toContain('msg=');
    });

    it('handles command injection in query values', () => {
      const cmd = encodeURIComponent('; rm -rf /');
      const result = ashCanonicalizeQueryNative(`cmd=${cmd}`);
      expect(result).toContain('cmd=');
    });

    it('handles path traversal in query values', () => {
      const traversal = encodeURIComponent('../../../etc/passwd');
      const result = ashCanonicalizeQueryNative(`file=${traversal}`);
      expect(result).toContain('file=');
    });

    it('handles extremely long keys', () => {
      const longKey = 'k'.repeat(1000);
      const result = ashCanonicalizeQueryNative(`${longKey}=value`);
      expect(result).toContain(`${longKey}=value`);
    });

    it('handles extremely long values', () => {
      const longValue = 'v'.repeat(10000);
      const result = ashCanonicalizeQueryNative(`key=${longValue}`);
      expect(result).toBe(`key=${longValue}`);
    });
  });

  describe('Native vs WASM Parity', () => {
    it('query canonicalization matches', () => {
      // Test cases without + which may be handled differently
      const testCases = [
        'a=1&b=2',
        'z=1&a=2&m=3',
        'a=%2f&b=%2F',
        '?a=1#frag',
        'empty=&key=value',
      ];

      for (const query of testCases) {
        expect(ashCanonicalizeQueryNative(query)).toBe(ashCanonicalizeQuery(query));
      }
    });

    it('handles plus sign - may differ between native and WASM', () => {
      // Native treats + as literal, WASM may encode it
      const result = ashCanonicalizeQueryNative('a+b=c+d');
      expect(result).toContain('a');
      expect(result).toContain('b');
    });

    it('urlencoded canonicalization matches', () => {
      const testCases = [
        'a=1&b=2',
      ];

      for (const form of testCases) {
        expect(ashCanonicalizeUrlencodedNative(form)).toBe(ashCanonicalizeUrlencoded(form));
      }
    });
  });

  describe('Real-World Query Patterns', () => {
    it('handles OAuth parameters', () => {
      const query = 'client_id=abc123&redirect_uri=https%3A%2F%2Fexample.com%2Fcallback&scope=read%20write&state=xyz789';
      const result = ashCanonicalizeQueryNative(query);
      expect(result).toContain('client_id=abc123');
      expect(result).toContain('redirect_uri=');
      expect(result).toContain('scope=');
      expect(result).toContain('state=xyz789');
    });

    it('handles search queries', () => {
      const query = 'q=hello%20world&page=1&limit=10&sort=relevance&filter%5Bcategory%5D=books';
      const result = ashCanonicalizeQueryNative(query);
      // Should be sorted alphabetically
      const parts = result.split('&');
      expect(parts[0]).toContain('filter');
      expect(parts[parts.length - 1]).toContain('sort');
    });

    it('handles API pagination', () => {
      const query = 'offset=100&limit=25&cursor=abc123&fields=id,name,email';
      const result = ashCanonicalizeQueryNative(query);
      expect(result).toBe('cursor=abc123&fields=id,name,email&limit=25&offset=100');
    });

    it('handles form submission data', () => {
      const form = 'first_name=John&last_name=Doe&email=john%40example.com&password=secret123';
      const result = ashCanonicalizeUrlencodedNative(form);
      expect(result).toContain('email=john%40example.com');
    });
  });
});

File: packages/ash-go/query_url_comprehensive_test.go
package ash

import (
	"fmt"
	"strings"
	"testing"
)

// ============================================================================
// COMPREHENSIVE QUERY AND URL ENCODING TESTS
// ============================================================================

// --- Basic Query Canonicalization ---

func TestQueryCanonicalizeEmpty(t *testing.T) {
	result, err := CanonicalizeQuery("")
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result != "" {
		t.Errorf("Empty query should return empty, got %s", result)
	}
}

func TestQueryCanonicalizeSingleParam(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1")
	if result != "a=1" {
		t.Errorf("Expected a=1, got %s", result)
	}
}

func TestQueryCanonicalizeMultipleParams(t *testing.T) {
	result, _ := CanonicalizeQuery("b=2&a=1")
	if result != "a=1&b=2" {
		t.Errorf("Expected sorted params a=1&b=2, got %s", result)
	}
}

func TestQueryCanonicalizeManyParams(t *testing.T) {
	result, _ := CanonicalizeQuery("z=1&y=2&x=3&w=4&v=5")
	if result != "v=5&w=4&x=3&y=2&z=1" {
		t.Errorf("Expected sorted params, got %s", result)
	}
}

// --- Duplicate Keys ---

func TestQueryDuplicateKeysSortedByValue(t *testing.T) {
	testCases := []struct {
		input    string
		expected string
	}{
		{"a=2&a=1", "a=1&a=2"},
		{"a=c&a=b&a=a", "a=a&a=b&a=c"},
		{"x=3&x=1&x=2", "x=1&x=2&x=3"},
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			result, _ := CanonicalizeQuery(tc.input)
			if result != tc.expected {
				t.Errorf("Expected %s, got %s", tc.expected, result)
			}
		})
	}
}

func TestQueryDuplicateMixedKeys(t *testing.T) {
	result, _ := CanonicalizeQuery("b=1&a=2&b=0&a=1")
	// Should sort by key first, then by value for same keys
	if !strings.Contains(result, "a=1") || !strings.Contains(result, "a=2") {
		t.Errorf("Should contain both a values, got %s", result)
	}
	if !strings.Contains(result, "b=0") || !strings.Contains(result, "b=1") {
		t.Errorf("Should contain both b values, got %s", result)
	}
}

// --- Empty Values ---

func TestQueryEmptyValueComprehensive(t *testing.T) {
	result, _ := CanonicalizeQuery("key=")
	if result != "key=" {
		t.Errorf("Expected key=, got %s", result)
	}
}

func TestQueryNoEquals(t *testing.T) {
	result, _ := CanonicalizeQuery("key")
	// Should handle key without value
	if result == "" && result != "key" && result != "key=" {
		t.Logf("Key without equals: %s", result)
	}
}

func TestQueryMixedEmptyValues(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1&b=&c=3")
	if !strings.Contains(result, "b=") {
		t.Errorf("Should preserve empty value, got %s", result)
	}
}

// --- Percent Encoding ---

func TestQueryPercentEncodingUppercaseComprehensive(t *testing.T) {
	testCases := []struct {
		input    string
		contains string
	}{
		{"a=%20", "%20"},      // Should keep uppercase
		{"a=%2f", "%2F"},      // Should uppercase
		{"a=%2F", "%2F"},      // Already uppercase
		{"a=%3a%3b", "%3A%3B"}, // Should uppercase both
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			result, _ := CanonicalizeQuery(tc.input)
			if !strings.Contains(result, tc.contains) {
				t.Errorf("Expected %s to contain %s, got %s", tc.input, tc.contains, result)
			}
		})
	}
}

func TestQuerySpaceEncoding(t *testing.T) {
	result, _ := CanonicalizeQuery("a=hello%20world")
	if !strings.Contains(result, "%20") {
		t.Errorf("Space should be encoded as %%20, got %s", result)
	}
}

func TestQueryPlusSign(t *testing.T) {
	result, _ := CanonicalizeQuery("a=hello+world")
	// Plus should be treated literally or converted to %2B
	if result == "" {
		t.Error("Result should not be empty")
	}
}

// --- Special Characters ---

func TestQuerySpecialCharacters(t *testing.T) {
	chars := []string{
		"a=!",
		"a=@",
		"a=#",
		"a=$",
		"a=&b=1",
		"a=*",
		"a=(",
		"a=)",
	}

	for _, input := range chars {
		t.Run(input, func(t *testing.T) {
			result, _ := CanonicalizeQuery(input)
			if result == "" && !strings.Contains(input, "&") {
				t.Error("Result should not be empty")
			}
		})
	}
}

// --- Leading Question Mark ---

func TestQueryLeadingQuestionMark(t *testing.T) {
	result, _ := CanonicalizeQuery("?a=1")
	// Should strip leading ?
	if strings.HasPrefix(result, "?") {
		t.Errorf("Should strip leading ?, got %s", result)
	}
	if !strings.Contains(result, "a=1") {
		t.Errorf("Should contain a=1, got %s", result)
	}
}

// --- Fragment Handling ---

func TestQueryFragmentStripped(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1#section")
	if strings.Contains(result, "#") {
		t.Errorf("Should strip fragment, got %s", result)
	}
}

// --- URL Encoded Canonicalization ---

func TestURLEncodedEmpty(t *testing.T) {
	result, err := CanonicalizeURLEncoded("")
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result != "" {
		t.Errorf("Expected empty, got %s", result)
	}
}

func TestURLEncodedSinglePair(t *testing.T) {
	result, _ := CanonicalizeURLEncoded("key=value")
	if result != "key=value" {
		t.Errorf("Expected key=value, got %s", result)
	}
}

func TestURLEncodedSorted(t *testing.T) {
	result, _ := CanonicalizeURLEncoded("z=3&a=1&m=2")
	if result != "a=1&m=2&z=3" {
		t.Errorf("Expected sorted, got %s", result)
	}
}

func TestURLEncodedDuplicates(t *testing.T) {
	result, _ := CanonicalizeURLEncoded("a=2&a=1")
	if result != "a=1&a=2" {
		t.Errorf("Expected a=1&a=2, got %s", result)
	}
}

func TestURLEncodedPlusAsLiteral(t *testing.T) {
	result, _ := CanonicalizeURLEncoded("a=hello+world")
	// Plus should be treated as literal
	if !strings.Contains(result, "+") && !strings.Contains(result, "%2B") {
		t.Logf("Plus handling: %s", result)
	}
}

// --- Case Sensitivity ---

func TestQueryCaseSensitiveKeysComprehensive(t *testing.T) {
	result, _ := CanonicalizeQuery("A=1&a=2")
	// A and a should be treated as different keys
	if !strings.Contains(result, "A=") || !strings.Contains(result, "a=") {
		t.Errorf("Keys should be case-sensitive, got %s", result)
	}
}

func TestQueryCaseSensitiveValues(t *testing.T) {
	result1, _ := CanonicalizeQuery("a=ABC")
	result2, _ := CanonicalizeQuery("a=abc")
	if result1 == result2 {
		t.Error("Values should be case-sensitive")
	}
}

// --- Unicode in Query ---

func TestQueryUnicodeValues(t *testing.T) {
	testCases := []string{
		"name=日本語",
		"emoji=🚀",
		"mixed=hello世界",
	}

	for _, input := range testCases {
		t.Run(input, func(t *testing.T) {
			result, err := CanonicalizeQuery(input)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			if result == "" {
				t.Error("Result should not be empty")
			}
		})
	}
}

func TestQueryUnicodeKeys(t *testing.T) {
	result, err := CanonicalizeQuery("キー=value")
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Unicode key should be handled")
	}
}

// --- Large Queries ---

func TestQueryLargeNumberOfParams(t *testing.T) {
	var params []string
	for i := 0; i < 100; i++ {
		params = append(params, fmt.Sprintf("key%03d=value%d", i, i))
	}
	input := strings.Join(params, "&")

	result, err := CanonicalizeQuery(input)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Result should not be empty")
	}

	// Verify sorted
	if !strings.HasPrefix(result, "key000=") {
		t.Error("Should be sorted alphabetically")
	}
}

func TestQueryLongValue(t *testing.T) {
	longValue := strings.Repeat("x", 10000)
	input := "key=" + longValue

	result, err := CanonicalizeQuery(input)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if !strings.Contains(result, longValue) {
		t.Error("Should preserve long value")
	}
}

// --- Determinism ---

func TestQueryDeterminism(t *testing.T) {
	inputs := []string{
		"z=1&y=2&x=3",
		"a=1&a=2&a=3",
		"key=value%20with%20spaces",
	}

	for _, input := range inputs {
		t.Run(input, func(t *testing.T) {
			result1, _ := CanonicalizeQuery(input)
			for i := 0; i < 100; i++ {
				result2, _ := CanonicalizeQuery(input)
				if result1 != result2 {
					t.Error("Query canonicalization should be deterministic")
				}
			}
		})
	}
}

func TestURLEncodedDeterminism(t *testing.T) {
	inputs := []string{
		"z=1&y=2&x=3",
		"a=1&a=2&a=3",
	}

	for _, input := range inputs {
		t.Run(input, func(t *testing.T) {
			result1, _ := CanonicalizeURLEncoded(input)
			for i := 0; i < 100; i++ {
				result2, _ := CanonicalizeURLEncoded(input)
				if result1 != result2 {
					t.Error("URL encoded canonicalization should be deterministic")
				}
			}
		})
	}
}

// --- Edge Cases ---

func TestQueryOnlyAmpersands(t *testing.T) {
	result, _ := CanonicalizeQuery("&&&")
	// Should handle gracefully
	if strings.Contains(result, "&&&") {
		t.Error("Should not preserve empty segments")
	}
}

func TestQueryTrailingAmpersand(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1&")
	// Should handle gracefully
	if strings.HasSuffix(result, "&") && !strings.HasSuffix(result, "a=1&") {
		t.Logf("Trailing ampersand handling: %s", result)
	}
}

func TestQueryLeadingAmpersand(t *testing.T) {
	result, _ := CanonicalizeQuery("&a=1")
	// Should handle gracefully
	if strings.HasPrefix(result, "&") {
		t.Error("Should not have leading ampersand")
	}
}

func TestQueryDoubleAmpersand(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1&&b=2")
	// Should handle gracefully
	if strings.Contains(result, "&&") {
		t.Error("Should not have double ampersand")
	}
}

func TestQueryDoubleEquals(t *testing.T) {
	result, _ := CanonicalizeQuery("a==1")
	// Value should be =1
	if result == "" {
		t.Error("Should handle double equals")
	}
}

func TestQueryEncodedAmpersand(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1%262")
	// %26 is encoded &, should be preserved in value
	if !strings.Contains(result, "%26") {
		t.Logf("Encoded ampersand: %s", result)
	}
}

func TestQueryEncodedEquals(t *testing.T) {
	result, _ := CanonicalizeQuery("a=1%3D2")
	// %3D is encoded =, should be preserved in value
	if !strings.Contains(result, "%3D") {
		t.Logf("Encoded equals: %s", result)
	}
}

// --- Real-World Query Strings ---

func TestQueryRealWorldSearch(t *testing.T) {
	query := "q=search+term&page=1&limit=10&sort=date&order=desc"
	result, err := CanonicalizeQuery(query)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Should handle real-world search query")
	}
}

func TestQueryRealWorldFilter(t *testing.T) {
	query := "filter[status]=active&filter[type]=user&filter[created_at][gte]=2024-01-01"
	result, err := CanonicalizeQuery(query)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Should handle filter query")
	}
}

func TestQueryRealWorldArrayParams(t *testing.T) {
	query := "ids[]=1&ids[]=2&ids[]=3"
	result, err := CanonicalizeQuery(query)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Should handle array params")
	}
}

func TestQueryRealWorldOAuth(t *testing.T) {
	query := "client_id=abc123&redirect_uri=https%3A%2F%2Fexample.com%2Fcallback&response_type=code&scope=read+write"
	result, err := CanonicalizeQuery(query)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if result == "" {
		t.Error("Should handle OAuth query")
	}
}

// --- Base64URL Tests ---

func TestBase64URLEncodeEmpty(t *testing.T) {
	result := Base64URLEncode([]byte{})
	if result != "" {
		t.Errorf("Empty input should return empty, got %s", result)
	}
}

func TestBase64URLEncodeBasic(t *testing.T) {
	result := Base64URLEncode([]byte("hello"))
	if result == "" {
		t.Error("Should encode basic string")
	}
	// Should not contain + or /
	if strings.Contains(result, "+") || strings.Contains(result, "/") {
		t.Error("Base64URL should not contain + or /")
	}
}

func TestBase64URLEncodePadding(t *testing.T) {
	result := Base64URLEncode([]byte("a"))
	// Should not have trailing =
	if strings.HasSuffix(result, "=") {
		t.Error("Base64URL should not have padding")
	}
}

func TestBase64URLDecodeEmpty(t *testing.T) {
	result, err := Base64URLDecode("")
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if len(result) != 0 {
		t.Error("Empty input should return empty")
	}
}

func TestBase64URLDecodeBasic(t *testing.T) {
	encoded := Base64URLEncode([]byte("hello"))
	decoded, err := Base64URLDecode(encoded)
	if err != nil {
		t.Fatalf("Unexpected error: %v", err)
	}
	if string(decoded) != "hello" {
		t.Errorf("Expected hello, got %s", string(decoded))
	}
}

func TestBase64URLRoundTrip(t *testing.T) {
	inputs := [][]byte{
		[]byte(""),
		[]byte("a"),
		[]byte("ab"),
		[]byte("abc"),
		[]byte("abcd"),
		[]byte("hello world"),
		[]byte{0x00, 0x01, 0x02, 0xFF, 0xFE},
	}

	for i, input := range inputs {
		t.Run(fmt.Sprintf("case_%d", i), func(t *testing.T) {
			encoded := Base64URLEncode(input)
			decoded, err := Base64URLDecode(encoded)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			if string(decoded) != string(input) {
				t.Errorf("Round trip failed: %v != %v", decoded, input)
			}
		})
	}
}
